src/v/storage/disk_log_impl.cc

```diff
@@ -519,10 +513,9 @@ ss::future<> disk_log_impl::adjacent_merge_compact(
         if (result.did_compact()) {
             segment->clear_cached_disk_usage();
             _compaction_ratio.update(result.compaction_ratio());
-            const auto removed_bytes = result.size_before - result.size_after;
-            subtract_dirty_segment_bytes(removed_bytes);
-            subtract_closed_segment_bytes(removed_bytes);
-
+            const ssize_t removed_bytes = ssize_t(result.size_before)
+                                          - ssize_t(result.size_after);
```

Comment:
Maybe the primitive should be `replace_segment_bytes(segment, uint64_t size_before, uint64_t size_after)`? Then I'm not sure we'd even need add/subtract primitives in disk_log_impl (adding/subtracting is then just a matter of supplying 0 for one of the args), and we wouldn't need to worry about signedness at this layer